beacon-chain/core/blocks/block_operations.go

```diff
@@ -524,6 +527,136 @@ func verifyAttestation(beaconState *pb.BeaconState, att *pb.Attestation) error {
+//     )
+func ProcessBlockValidatorDeposits(
+	beaconState *pb.BeaconState,
+	block *pb.BeaconBlock,
+) (*pb.BeaconState, error) {
+	deposits := block.GetBody().GetDeposits()
+	if uint64(len(deposits)) > params.BeaconConfig().MaxDeposits {
+		return nil, fmt.Errorf(
+			"number of deposits (%d) exceeds allowed threshold of %d",
+			len(deposits),
+			params.BeaconConfig().MaxDeposits,
+		)
+	}
+	var err error
+	var depositInput *pb.DepositInput
+	for idx, deposit := range deposits {
+		depositData := deposit.GetDepositData()
+		depositInput, err = DecodeDepositInput(depositData)
+		if err != nil {
+			return nil, fmt.Errorf("could not decode deposit input: %v", err)
+		}
+		if err = verifyDeposit(beaconState, deposit); err != nil {
+			return nil, fmt.Errorf("could not verify deposit #%d: %v", idx, err)
+		}
+		// depositData consists of depositInput []byte + depositValue [8]byte +
+		// depositTimestamp [8]byte.
+		depositValue := depositData[len(depositData)-16 : len(depositData)-8]
+		// We then mutate the beacon state with the verified validator deposit.
+		beaconState, _, err = v.ProcessDeposit(
+			beaconState,
+			depositInput.GetPubkey(),
+			binary.BigEndian.Uint64(depositValue),
+			depositInput.GetProofOfPossession(),
+			depositInput.GetWithdrawalCredentialsHash32(),
+			depositInput.GetRandaoCommitmentHash32(),
+			depositInput.GetPocCommitment(),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could process deposit into beacon state: %v", err)
+		}
+	}
+	return beaconState, nil
+}
+
+// DecodeDepositInput unmarshales a depositData byte slice into
+// a proto *pb.DepositInput by using the Simple Serialize (SSZ)
+// algorithm.
+func DecodeDepositInput(depositData []byte) (*pb.DepositInput, error) {
+	// Last 16 bytes of deposit data are 8 bytes for value
+	// and 8 bytes for timestamp. Everything before that is a
+	// Simple Serialized deposit input value.
+	if len(depositData) < 16 {
+		return nil, fmt.Errorf(
+			"deposit data slice too small: len(depositData) = %d",
+			len(depositData),
+		)
+	}
+	depositInput := new(pb.DepositInput)
+	depositInputBytes := depositData[:len(depositData)-16]
+	rBuf := bytes.NewReader(depositInputBytes)
+	if err := ssz.Decode(rBuf, depositInput); err != nil {
+		return nil, fmt.Errorf("ssz decode failed: %v", err)
+	}
+	return depositInput, nil
+}
+
+func verifyDeposit(beaconState *pb.BeaconState, deposit *pb.Deposit) error {
+	depositData := deposit.GetDepositData()
+	// Verify Merkle proof of deposit and PoW receipt trie root.
+	var receiptRoot [32]byte
+	var merkleLeaf [32]byte
+	copy(receiptRoot[:], beaconState.GetProcessedPowReceiptRootHash32())
+	copy(merkleLeaf[:], depositData)
+	if ok := trie.VerifyMerkleBranch(
+		merkleLeaf,
+		deposit.GetMerkleBranchHash32S(),
+		params.BeaconConfig().DepositContractTreeDepth,
+		receiptRoot,
+	); !ok {
+		return errors.New("Deposit merkle branch of PoW receipt root did not verify")
```

Comment:
we should probably print out the `receiptRoot` here, later debugging can be a pain